### ISSUE TITLE
Recognize or provide directory path for compilation database files?

### ISSUE THREAD

**Issue by siegel** [clangd/clangd#732]
This may be more of a how-to than a feature request, if it already exists. Or it may be a request for a feature. :-)

Today I learned that `clang` has a command-line flag that will emit a compile database JSON fragment into a user-specified directory, corresponding to the compiled file. This is a *huge* leap forward for bridging the gap between Xcode projects and `compile_commands.json` without relying on external utilities, and it seems to work quite well with Xcode and `xcodebuild`.

Specifically, in my build configuration I specify:

`OTHER_CFLAGS = $(inherited) -gen-cdb-fragment-path "$(SRCROOT)/cdb"`

where `cdb` is a directory that I have created to hold the resulting files.

(More about `-gen-cdb-fragment-path` is [here](https://reviews.llvm.org/D66555) and it makes reference to an `-MJ` option, which my simple digging shows is handy for emitting a discrete JSON file for the compiled file rather than specifying a destination directory.)

After building, this directory contains a bunch of JSON files, one per compiled file though there may be multiples per individual source file, depending on whether dependent target(s) and build configurations need to compile the file more than once.

The results still need some postprocessing, but using `-gen-cdb-fragment-path` saves a lot of work, and as a bonus, the directory gets updated when build settings change or files get added in the Xcode project(though of course obsolete files probably don't go away if a file is removed from the build system; easy enough to rectify that by cleaning).

But this got me wondering: is it currently possible for `clangd` to recognize, or be told on a per-project basis, where a _folder_ full of compilation database fragments is located, and ingest the JSON files in the directory? If it's not, could it be?

Thanks for reading.

**Comment by kadircet** [comment]
> is it currently possible for clangd to recognize, or be told on a per-project basis, where a folder full of compilation database fragments is located, and ingest the JSON files in the directory?

Yes and no. Currently you can tell where your compilation database is via config:

```yaml
If:
  PathMatch: /path/to/project1/.*
CompileFlags:
  CompilationDatabase: /path/to/dir/with/cdb/project1/
---
If:
  PathMatch: /path/to/project2/.*
CompileFlags:
  CompilationDatabase: /path/to/dir/with/cdb/project2/
```

(you can also have a project local config and get rid of the `if/pathmatch` blocks, more details in https://clangd.llvm.org/config.html)

but clangd will only look for `compile_flags.txt` or `compile_commands.json` within that directory and will not perform merging for all the fragments with a given filename prefix.

> If it's not, could it be?

Maybe. We've got some caching/reloading logic in place for compilation databases, and parts of this logic resides in clang rather than clangd. So this will end up making things more complicated in multiple places, hence we might need some justifications.

You were talking about some post-processing anyways, any chance you could perform the merging as that post-processing step and put an auto-generated compile_commands.json in the directory you are going to tell clangd to look at ?

**Comment by siegel** [comment]
Thanks for the tip about the configuration info, I will ponder that.

> You were talking about some post-processing anyways, any chance you could perform the merging as that post-processing step and put an auto-generated compile_commands.json in the directory you are going to tell clangd to look at ?

"Answer hazy, ask later." :-) What I'm *really* trying to do here is to find a minimal recipe so that Xcode users (by far the majority of those using clangd on macOS) have a solution that allows them to have an enjoyable experience with clangd when using my client.

To that goal, the fewer (and the simpler) steps the Xcode user has to perform in order to get the necessary JSON data to clangd, the more likely it is that they'll perform those steps.

So, while *I* can certainly wire together a build script which would coalesce all of the CDB fragments (perhaps postprocessing as I go), it's a pretty steep hill for a lot of users to climb. (And any solution would be specific to my own setup.)

So I'm thinking that a combination of clangd being able to recognize a directory full of fragments, and in some cases my client being able to specify a pre-built config file which advises clangd to ignore certain options that Xcode generates (which cause clangd to malfunction or misbehave), would serve to smooth the path significantly for Xcode users.

**Comment by gsingh93** [comment]
I'm not sure if my usecase fits under this issue, but I have two compilation databases generated from building the Linux kernel. One is from a build with `ARCH=arm64` and the other is from a build with `ARCH=x86_64`. I'd love it if there was a way to merge the compilation databases so I could use `clangd` with architecture specific files from either target as well as shared files between the two targets.

In cases where I need to navigate to a definition but there's one available for both targets, having the ability to set the primary or default compilation database would be convenient.
